omniscidb/QueryEngine/CgenState.cpp

```diff
@@ -200,7 +200,10 @@ void CgenState::maybeCloneFunctionRecursive(llvm::Function* fn, bool is_l0) {
   for (auto it = llvm::inst_begin(fn), e = llvm::inst_end(fn); it != e; ++it) {
     if (llvm::isa<llvm::CallInst>(*it)) {
       auto& call = llvm::cast<llvm::CallInst>(*it);
-      maybeCloneFunctionRecursive(call.getCalledFunction(), is_l0);
+      // Ignore indirect calls (e.g. virtual function calls).
+      if (call.getCalledFunction()) {
+        maybeCloneFunctionRecursive(call.getCalledFunction(), is_l0);
```

Comment:
From LLVM docs: `getCalledFunction Returns the function called, or null if this is an indirect function invocation or the function signature does not match the call signature.`
So it's mostly the same as `isIndirectCall` but will also avoid SEGFAULT on wrong LLVM IR causing runtime exception instead later on LLVM check. Wrong LLVM appears from time to time in working branches when new features are added and having a proper error with a description is much more preferable.

Reply:
We have a check in the `maybeCloneFunctionRecursive` that catches exactly that the function is not present in the module, but expected to be. Adding the `if` masks the problem, so we will detect it only on IR validation (worst case, after linking). I don't see how it is more preferable.

Reply:
This is not the case. `maybeCloneFunctionRecursive` is not an IR integrity checker, it gets a declaration and checks if we have a definition of the function in addition to the declaration and copies this definition. At least for CPU, function definitions are optional in the runtime module (and we don't have them for quantile `*_impl` methods).

Introduced `if-statement` checks that we have a declaration for a call to be passed to `maybeCloneFunctionRecursive`. Otherwise, we would simply get a SEGFAULT. It would skip both indirect and incorrect calls, but it is OK because LLVM IR checkers would issue nice verbose errors about wrong calls, we don't need to construct our own checkers.

Reply:
It is. Function definitions are optional, yes. What is not optional is the presence of the LLVM function in the module that we clone from. It can be a declaration as in your `*_impl` example. A missing declaration or implementation is an internal error, a bug, so the function should fail on assertion (SEGFAULT). This is precisely what happens since the first thing `maybeCloneFunctionRecursive` does is check if the function is null. It does not have a nice description, but that's easily fixable.

Integrity checking is not enough. Later stages of code generation do not distinguish between a declaration of a `*_impl` function (without definition) we should have provided and an intrinsic declaration provided by a GPU driver. Even though the IR is "correct", PTX/SPIRV generation will fail. This leads to CPU fallback, masking a real bug. Cloning a function from the runtime module should not silently ignore missing functions.

Reply:
I don't see why you want to duplicate such checks. If a function is not present then you won't be able to emit call instruction. And existence of declarations for runtime module code is checked on runtime module compilation. And the problem with a function declaration not provided by GPU driver cannot be solved in this code anyway.

I'm OK with replacing the check with `isIndirectCall` though.